include/grpcpp/impl/codegen/server_callback_impl.h

```diff
@@ -353,835 +393,308 @@ class ServerBidiReactor : public internal::ServerReactor {
+template <class Request>
 class ServerReadReactor : public internal::ServerReactor {
  public:
+  ServerReadReactor() : reader_(nullptr) {}
   ~ServerReadReactor() = default;
 
   /// The following operation initiations are exactly like ServerBidiReactor.
-  void StartSendInitialMetadata() { reader_->SendInitialMetadata(); }
-  void StartRead(Request* req) { reader_->Read(req); }
-  void Finish(::grpc::Status s) { reader_->Finish(std::move(s)); }
-
-  /// Similar to ServerBidiReactor::OnStarted, except that this also provides
-  /// the response object that the stream fills in before calling Finish.
-  /// (It must be filled in if status is OK, but it may be filled in otherwise.)
-  ///
-  /// \param[in] context The context object now associated with this RPC
-  /// \param[in] resp The response object to be used by this RPC
-  virtual void OnStarted(::grpc_impl::ServerContext* /*context*/,
-                         Response* /*resp*/) {}
+  void StartSendInitialMetadata() {
+    auto* reader = reader_.load(std::memory_order_acquire);
+    if (reader == nullptr) {
+      grpc::internal::MutexLock l(&reader_mu_);
+      reader = reader_.load(std::memory_order_acquire);
+      if (reader == nullptr) {
+        send_initial_metadata_wanted_ = true;
+        return;
+      }
+    }
+    reader->SendInitialMetadata();
+  }
+  void StartRead(Request* req) {
+    auto* reader = reader_.load(std::memory_order_acquire);
+    if (reader == nullptr) {
+      grpc::internal::MutexLock l(&reader_mu_);
+      reader = reader_.load(std::memory_order_acquire);
+      if (reader == nullptr) {
+        read_wanted_ = req;
+        return;
+      }
+    }
+    reader->Read(req);
+  }
+  void Finish(::grpc::Status s) {
+    auto* reader = reader_.load(std::memory_order_acquire);
+    if (reader == nullptr) {
+      grpc::internal::MutexLock l(&reader_mu_);
+      reader = reader_.load(std::memory_order_acquire);
+      if (reader == nullptr) {
+        finish_wanted_ = true;
+        status_wanted_ = std::move(s);
+        return;
+      }
+    }
+    reader->Finish(std::move(s));
+  }
 
   /// The following notifications are exactly like ServerBidiReactor.
   virtual void OnSendInitialMetadataDone(bool /*ok*/) {}
   virtual void OnReadDone(bool /*ok*/) {}
-  void OnDone() override {}
+  void OnDone() override = 0;
   void OnCancel() override {}
 
  private:
   friend class ServerCallbackReader<Request>;
+
   // May be overridden by internal implementation details. This is not a public
   // customization point.
   virtual void InternalBindReader(ServerCallbackReader<Request>* reader) {
-    reader_ = reader;
+    grpc::internal::MutexLock l(&reader_mu_);
+    reader_.store(reader, std::memory_order_release);
+    if (send_initial_metadata_wanted_) {
+      reader->SendInitialMetadata();
+    }
+    if (read_wanted_ != nullptr) {
+      reader->Read(read_wanted_);
+    }
+    if (finish_wanted_) {
+      reader->Finish(std::move(status_wanted_));
+    }
```

Comment:
Is it worth resetting these fields after being processed once?

For example, if `read_wanted_`  is not `nullptr`, we should run Read and also set `read_wanted_` to `nullptr`?

(same question for the writer)

Reply:
Hypothetically (and it's hypothetical because we can't use C++17 right now here), could we use a std::variant between the `reader` and a struct containing the `_wanted`? I think that would give the strongest statement that the combination of these things is not relevant at the same time (though I'm not sure how `variant` plays with `atomic`). That said, we can reset values to nullptr to ease readability, I'm fine with that.

Reply:
Thank you for add the resets.  It just occurred to me to ask whether it's ok to call `Finish()`, `Read()` and `SendInitialMetadata()` while holding the lock? My question is mainly about performance.

We can potentially just get and reset the pointers and call the stream methods after Lock is released.

Reply:
Well, that's interesting. It's safe from a lock-inversion pov etc but yeah, there may be a performance impact especially for streaming cases. I'll take a crack at this right away.